email-reader/src/main/java/no/sikt/nva/email/reader/handler/ScopusEmailReader.java

```diff
@@ -4,46 +4,95 @@
+import no.sikt.nva.email.reader.service.FileRetriever;
+import no.sikt.nva.email.reader.service.ZipFileRetriever;
 import no.unit.nva.s3.S3Driver;
+import nva.commons.core.Environment;
+import nva.commons.core.JacocoGenerated;
+import nva.commons.core.paths.UnixPath;
 import nva.commons.core.paths.UriWrapper;
 import org.apache.james.mime4j.dom.Message;
 import software.amazon.awssdk.services.s3.S3Client;
-import nva.commons.core.JacocoGenerated;
 
+import java.io.InputStream;
 import java.net.URI;
 import java.net.http.HttpClient;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 
 public class ScopusEmailReader implements RequestHandler<S3Event, Set<URI>> {
 
 
+    public static final String UNABLE_TO_DOWNLOAD_FILE = "Unable to download file";
+    public static final String COULD_NOT_PERSIST_FILE_IN_S_3_BUCKET = "Could not persist file in s3 bucket";
     private static final String S3_URI_TEMPLATE = "s3://%s/%s";
     private static final int SINGLE_EXPECTED_RECORD = 0;
     private final S3Client s3Client;
 
-    @SuppressWarnings({"PMD.UnusedPrivateField", "PMD.SingularField"})
-    private final HttpClient httpClient; //TODO: implement downloading
+    private final FileRetriever fileRetriever;
+    private final String scopusZipBucket;
 
     @JacocoGenerated
     public ScopusEmailReader() {
         this(S3Driver.defaultS3Client().build(),
-                HttpClient.newBuilder().build());
+                new ZipFileRetriever(HttpClient.newBuilder().build()),
+                new Environment().readEnv("SCOPUS_ZIP_BUCKET"));
     }
 
-    public ScopusEmailReader(S3Client s3Client, HttpClient httpClient) {
+    public ScopusEmailReader(S3Client s3Client, FileRetriever fileRetriever, String scopusZipBucket) {
         this.s3Client = s3Client;
-        this.httpClient = httpClient;
+        this.fileRetriever = fileRetriever;
+        this.scopusZipBucket = scopusZipBucket;
     }
 
     @Override
     public Set<URI> handleRequest(S3Event event, Context context) {
-        var emailString = getEmailFromS3(event);
-        var message = extractMessage(event, emailString);
-        return extractUrisFromMessage(event, message);
+        return Try.of(() -> getEmailFromS3(event))
+                .mapTry(email -> extractMessage(event, email))
+                .mapTry(message -> extractUrisFromMessage(event, message))
+                .mapTry(uriSet -> downloadZipFilesAndPersistThem(uriSet, event))
+                .getOrElseThrow(throwable -> handleFailure(throwable, event));
+    }
+
+    private Set<URI> downloadZipFilesAndPersistThem(Set<URI> uris, S3Event event) {
+        var s3Driver = new S3Driver(s3Client, scopusZipBucket);
+        return uris.stream().map(uri -> putOnS3(uri, s3Driver, event)).collect(Collectors.toSet());
+    }
+
+    private URI putOnS3(URI uri, S3Driver s3Driver, S3Event event) {
+        var objectKey = UnixPath.of(UriWrapper.fromUri(uri).getLastPathElement());
```

Comment:
Is this name unique? Do we rely on it not being unique? Might be tempted to explain this in a method name in case someone misunderstands which of these is being assumed.

Reply:
I wasn't to sure about  whether to craft a sortable-identifier based name with a ".zip" ending. I was peeking into what Nikolai had previously done, and well since the last path element contains the date, it might not be such a stupid idea to preserve it.

Reply:
Not only contains the date but also what sort of content is has. Ex. full abstracs or delete list.

Reply:
Checked how random keys here would work:

We are depending on the names from the scopus dataset eventually to be object-keys (as scopus-identifiers)

We could go through the whole system and give random names all over but that seems a bit over-engineered.

But I don't see how that would benefit us, other than making monitoring more complicated.